Need reliable health check for SharedInformer

We are using kubernetes-client version 5.2.1.

We observed an issue where our SharedInformer cache does not keep up with a new resource created in kubernetes. In this scenario the SharedInformer initially synced correctly and was running for a few days. The issue appears to be originating in the cluster and possibly the cluster is somehow overwhelmed although the system is not under heavy load. It has only happened in a live customer environment.

The issue appears to be triggered after the creation in kubernetes of a resource of the same type as the SharedInformer. We see a socket/connection exception in the logs that appears to be related to the SharedInformer watch connection. The exception is happening in okhttp thread and being logged by fabric8 WatcherWebSocketListener.

The main issue is that the SharedInformer continues to function with stale data even after the exception, i.e. the SharedInformer cache does not get updated with the newly created resource. In fact we have code in our application that calls the hasSynced method on each access of the SharedInformer cache. It appears that the hasSynced must be returning true because we have observed a different scenario where the informer fails to sync and in that scenario our application fails.

It has been mentioned that SharedInformer has an isRunning method but it's not clear that that method is intended as a reliable health check.

We would like to have a reliable health check for SharedInformer that would detect this exception.

```
Caused by: javax.net.ssl.SSLException: Socket closed
        at sun.security.ssl.Alert.createSSLException(Alert.java:127)
        at sun.security.ssl.TransportContext.fatal(TransportContext.java:353)
        at sun.security.ssl.TransportContext.fatal(TransportContext.java:296)
        at sun.security.ssl.TransportContext.fatal(TransportContext.java:291)
        at sun.security.ssl.SSLSocketImpl.handleException(SSLSocketImpl.java:1522)
        at sun.security.ssl.SSLSocketImpl$AppInputStream.read(SSLSocketImpl.java:965)
        at okio.Okio$2.read(Okio.java:140)
        at okio.AsyncTimeout$2.read(AsyncTimeout.java:237)
        ... 38 more
Caused by: java.net.SocketException: Socket closed
        at java.net.SocketInputStream.read(SocketInputStream.java:183)
        at java.net.SocketInputStream.read(SocketInputStream.java:140)
        at sun.security.ssl.SSLSocketInputRecord.read(SSLSocketInputRecord.java:476)
        at sun.security.ssl.SSLSocketInputRecord.readHeader(SSLSocketInputRecord.java:470)
        at sun.security.ssl.SSLSocketInputRecord.bytesInCompletePacket(SSLSocketInputRecord.java:70)
        at sun.security.ssl.SSLSocketImpl.readApplicationRecord(SSLSocketImpl.java:1308)
        at sun.security.ssl.SSLSocketImpl$AppInputStream.read(SSLSocketImpl.java:949)
        ... 40 more
2021-05-05 14:52:03.621+0000 |  | ::: | OkHttp https://10.96 |  | ERROR | io.fabric8.kubernetes.client.informers.cache.ReflectorWatcher | Watch closing
2021-05-05 16:02:03.540+0000 |  | ::: | OkHttp https://10.96 |  | ERROR | io.fabric8.kubernetes.client.informers.cache.ReflectorWatcher | Watch closing
2021-05-05 16:07:08.656+0000 |  | ::: | OkHttp WebSocket htt |  |  WARN | io.fabric8.kubernetes.client.dsl.internal.WatcherWebSocketListener | Exec Failure
java.net.SocketTimeoutException: sent ping but didn't receive pong within 30000ms (after 8 successful ping/pongs)
        at okhttp3.internal.ws.RealWebSocket.writePingFrame(RealWebSocket.java:545)
        at okhttp3.internal.ws.RealWebSocket$PingRunnable.run(RealWebSocket.java:529)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.lang.Thread.run(Thread.java:834)
        at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:517)
        at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:192)
2021-05-05 16:07:09.723+0000 |  | ::: | OkHttp https://10.96 |  | ERROR | io.fabric8.kubernetes.client.informers.cache.ReflectorWatcher | Watch closing
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -29,6 +29,7 @@
 * Fix #3186: WebSockets and HTTP connections are closed as soon as possible for Watches.
 * Fix #2937: Add `SharedInformerFactory#getExistingSharedIndexInformers` method to return list of registered informers
 * Fix #3239: Add the `Informable` interface for context specific dsl methods to create `SharedIndexInformer`s.
+* Fix #3101: making isWatching a health check for the informer
 
 #### Dependency Upgrade
 * Fix #2741: Update Knative Model to v0.23.0

--- a/doc/CHEATSHEET.md
+++ b/doc/CHEATSHEET.md
@@ -1831,7 +1831,7 @@ Kubernetes Client also provides `SharedInformer` support in order to stay update
 ```java
 SharedInformerFactory sharedInformerFactory = client.informers();
 ```
-- Create `SharedIndexInformer` for some Kubernetes Resource(requires resource's class and resync period(when to check with server again while watching something).  By default it watches in all namespaces.:
+- Create `SharedIndexInformer` for some Kubernetes Resource(requires resource's class and resync period (emits a dummy update event on that interval so that the handler can act again).  By default it watches in all namespaces.:
 ```java
 SharedIndexInformer<Pod> podInformer = sharedInformerFactory.sharedIndexInformerFor(Pod.class, 30 * 1000L);
 podInformer.addEventHandler(new ResourceEventHandler<Pod>() {
@@ -1871,15 +1871,20 @@ dummyInformer.addEventHandler(new ResourceEventHandler<Dummy>() {
   }
 });
 ```
+- Start all registered informers:
+```java
+sharedInformerFactory.startAllRegisteredInformers();
+```
+- Stop all registered informers:
+```java
+sharedInformerFactory.stopAllRegisteredInformers();
+```
+
+You are not limited to just creating cluster wide informers, if you want to be informed about a particular context then use the Informable interface and inform methods.
+
 - Create namespaced `SharedIndexInformer` (informers specific to a particular `Namespace`):
 ```java
-SharedInformerFactory sharedInformerFactory = client.informers();
-SharedIndexInformer<Pod> podInformer = sharedInformerFactory.inNamespace("default").sharedIndexInformerFor(
-        Pod.class,
-        30 * 1000L);
-logger.info("Informer factory initialized.");
-
-podInformer.addEventHandler(new ResourceEventHandler<Pod>() {
+SharedIndexInformer<Pod> podInformer = client.pods().inNamespace("default").inform(new ResourceEventHandler<Pod>() {
     @Override
     public void onAdd(Pod pod) {
         logger.info("Pod " + pod.getMetadata().getName() + " got added");
@@ -1894,8 +1899,9 @@ podInformer.addEventHandler(new ResourceEventHandler<Pod>() {
     public void onDelete(Pod pod, boolean deletedFinalStateUnknown) {
         logger.info("Pod " + pod.getMetadata().getName() + " got deleted");
     }
-});
-}
+},  30 * 1000L);
+
+logger.info("Informer initialized.");
 ```
 - Create Namespaced Informer for a Custom Resource(**Note:** Your CustomResource POJO must implement `Namespaced` interface like the one used in this example: [Dummy.java](https://github.com/fabric8io/kubernetes-client/blob/master/kubernetes-examples/src/main/java/io/fabric8/kubernetes/examples/crds/Dummy.java))
 You should have your CustomResource type POJO annotated with group, version fields with respect to your CRD:
@@ -1913,33 +1919,24 @@ public class Dummy extends CustomResource<DummySpec, KubernetesResource> impleme
 ```
 Then you should be able to use it like this:
 ```java
-SharedIndexInformer<Dummy> dummyInformer = sharedInformerFactory.inNamespace("default").sharedIndexInformerForCustomResource(Dummy.class, 60 * 1000L);
-dummyInformer.addEventHandler(new ResourceEventHandler<Dummy>() {
-  @Override
-  public void onAdd(Dummy dummy) {
-    System.out.printf("%s dummy added\n", dummy.getMetadata().getName());
-  }
+SharedIndexInformer<Dummy> dummyInformer = client.customResources(Dummy.class).inNamespace("default").inform(new ResourceEventHandler<Dummy>() {
+    @Override
+    public void onAdd(Dummy dummy) {
+        System.out.printf("%s dummy added\n", dummy.getMetadata().getName());
+    }
 
-  @Override
-  public void onUpdate(Dummy oldDummy, Dummy newDummy) {
-    System.out.printf("%s dummy updated\n", oldDummy.getMetadata().getName());
-  }
+    @Override
+    public void onUpdate(Dummy oldDummy, Dummy newDummy) {
+        System.out.printf("%s dummy updated\n", oldDummy.getMetadata().getName());
+    }
 
-  @Override
-  public void onDelete(Dummy dummy, boolean deletedFinalStateUnknown) {
-    System.out.printf("%s dummy deleted \n", dummy.getMetadata().getName());
-  }
-});
+    @Override
+    public void onDelete(Dummy dummy, boolean deletedFinalStateUnknown) {
+        System.out.printf("%s dummy deleted \n", dummy.getMetadata().getName());
+    }
+},  60 * 1000L);
 ```
-
-- Start all registered informers:
-```java
-sharedInformerFactory.startAllRegisteredInformers();
-```
-- Stop all registered informers:
-```java
-sharedInformerFactory.stopAllRegisteredInformers();
-```
+When using the inform methods the informers will already be started/running.
 
 ### List Options
 There are various options provided by Kubernetes Client API when it comes to listing resources. Here are some of the common examples provided:

--- a/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/informers/cache/Reflector.java
+++ b/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/informers/cache/Reflector.java
@@ -81,7 +81,6 @@ public class Reflector<T extends HasMetadata, L extends KubernetesResourceList<T
    */
   public void listSyncAndWatch() {
     running = true;
-    watching = false;
     final L list = getList();
     final String latestResourceVersion = list.getMetadata().getResourceVersion();
     lastSyncResourceVersion = latestResourceVersion;
@@ -155,12 +154,21 @@ public class Reflector<T extends HasMetadata, L extends KubernetesResourceList<T
 
     @Override
     public void onClose(WatcherException exception) {
-      watchStopped();
       // this close was triggered by an exception,
       // not the user, it is expected that the watch retry will handle this
       log.warn("Watch closing with exception", exception);
-      if (exception.isHttpGone()) {
+      boolean restarted = false;
+      try {
+        if (exception.isHttpGone()) {
           listSyncAndWatch();
+          restarted = true;
+        } else {
+          running = false; // shouldn't happen, but it means the watch won't restart
+        }
+      } finally {
+        if (!restarted) {
+          watchStopped(); // report the watch as stopped after a problem
+        }
       }
     }
 
@@ -175,6 +183,10 @@ public class Reflector<T extends HasMetadata, L extends KubernetesResourceList<T
       return true;
     }
     
+  }
+  
+  public ReflectorWatcher getWatcher() {
+    return watcher;
   }
 
 }

--- a/kubernetes-client/src/test/java/io/fabric8/kubernetes/client/informers/cache/ReflectorTest.java
+++ b/kubernetes-client/src/test/java/io/fabric8/kubernetes/client/informers/cache/ReflectorTest.java
@@ -21,6 +21,7 @@ import io.fabric8.kubernetes.api.model.PodList;
 import io.fabric8.kubernetes.api.model.PodListBuilder;
 import io.fabric8.kubernetes.client.KubernetesClientException;
 import io.fabric8.kubernetes.client.Watch;
+import io.fabric8.kubernetes.client.WatcherException;
 import io.fabric8.kubernetes.client.dsl.base.OperationContext;
 import io.fabric8.kubernetes.client.informers.ListerWatcher;
 import org.junit.jupiter.api.Test;
@@ -29,7 +30,6 @@ import org.mockito.Mockito;
 import static org.junit.Assert.assertTrue;
 import static org.junit.jupiter.api.Assertions.assertFalse;
 import static org.junit.jupiter.api.Assertions.assertThrows;
-import static org.junit.jupiter.api.Assertions.fail;
 
 class ReflectorTest {
 
@@ -62,6 +62,26 @@ class ReflectorTest {
     assertTrue(reflector.isRunning());
 
     reflector.stop();
+
+    assertFalse(reflector.isWatching());
+    assertFalse(reflector.isRunning());
+  }
+  
+  @Test
+  void testNonHttpGone() {
+    ListerWatcher<Pod, PodList> mock = Mockito.mock(ListerWatcher.class);
+    PodList list = new PodListBuilder().withNewMetadata().withResourceVersion("1").endMetadata().build();
+    Mockito.when(mock.list(Mockito.any(), Mockito.any(), Mockito.any())).thenReturn(list);
+
+    Reflector<Pod, PodList> reflector =
+        new Reflector<>(Pod.class, mock, Mockito.mock(SyncableStore.class), new OperationContext());
+
+    reflector.listSyncAndWatch();
+
+    assertTrue(reflector.isWatching());
+    assertTrue(reflector.isRunning());
+
+    reflector.getWatcher().onClose(new WatcherException(null));
 
     assertFalse(reflector.isWatching());
     assertFalse(reflector.isRunning());

